### ISSUE TITLE
Should well-known identifiers be used for ubiquitous payment methods

### ISSUE THREAD

**Issue by adrianba** [WICG/paymentrequest#35]
The Payment Method Identifiers spec calls out some open issues. Payment method identifiers are strings that developers need to include in their code to indicate that a merchant can accept a payment method. There may be popular payment methods that will be commonly typed for which we don't want to require the full distributed extensibility format. If we choose to support these then we also need to decide where they are specified.

**Comment by adrianhopebailie** [comment]
Is there any value in having payment method identifiers that are not URIs?

**Comment by ianbjacobs** [comment]
I think many developers consider URIs suboptimal identifiers (too long, hard to remember, etc.). 

From an end user perspective it's not an issue as they are unlikely to ever see these identifiers.

**Comment by adrianba** [comment]
Some of the problems are the same as XML Namespaces. Some points described [here](https://wiki.whatwg.org/wiki/Why_no_namespaces) are relevant. Ian notes some (long, hard to remember) but also inconsistent (end in a slash? casing? https or http? include www.?).

I am reluctantly coming to terms with the fact that we will need to use something that either looks like a URL or that can be consistently mapped to a URL for #34 and in my mind this provides a stronger case for identifiers that are not a URL (which I had originally wanted to avoid).

**Comment by webpayments** [comment]
I personally like short identifiers that map to URLs. And of course some
sort of registry for same ;-)

On Thu, Dec 10, 2015 at 10:00 AM, Ade Bateman <notifications@github.com>
wrote:

> Some of the problems are the same as XML Namespaces. Some points described
> here <https://wiki.whatwg.org/wiki/Why_no_namespaces> are relevant. Ian
> notes some (long, hard to remember) but also inconsistent (end in a slash?
> casing? https or http? include www.?).
>
> I am reluctantly coming to terms with the fact that we will need to use
> something that either looks like a URL or that can be consistently mapped
> to a URL for #34 <https://github.com/WICG/paymentrequest/issues/34> and
> in my mind this provides a stronger case for identifiers that are not a URL
> (which I had originally wanted to avoid).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/WICG/paymentrequest/issues/35#issuecomment-163669875>.
>



-- 
-Shane

**Comment by adrianhopebailie** [comment]
> Some of the problems are the same as XML Namespaces. Some points described here are relevant.

We are talking about identifiers not namespaces so I think a lot of the reasons in that reference are irrelevant.

At this stage we're not expecting developers to have to namespace the data in their payment requests or responses so there's no risk that these URLs become anything more than identifiers.

A payment method identifier needs to be globally unique and it should be possible for anyone to mint one easily. 

Also, as @ianbjacobs says end user's will never see these so how pretty or easy to remember they are is irrelevant to my mind.

We should provide guidance up front around conventions such as http vs https and the presence or not of trailing slashes to help with consistency.

> I personally like short identifiers that map to URLs. And of course some sort of registry for same ;-)

Why add the overhead of a registry if it's not actually required?

**Comment by adrianba** [comment]
> We are talking about identifiers not namespaces so I think a lot of the reasons in that reference are irrelevant.

I didn't say all, I said some. And I still think many of them are relevant.

>how pretty or easy to remember they are is irrelevant to my mind.

We should keep in mind the [priority of constituencies](http://www.w3.org/TR/html-design-principles/#priority-of-constituencies). While users won't see these, potentially millions of developers will. Everytime somebody builds a web page that calls the API they will need to type these identifiers.

**Comment by adrianhopebailie** [comment]
This is starting to feel a lot like a duplicate of #34, should we close one of them?

> And I still think many of them are relevant.

The only ones that I would say are relevant to URI's as identifiers we are already discussing and essentially the argument against them boils down to: _"developers are lazy and so on the odd occasion when they will hand-code a payment request we shouldn't make them write out whole URLs"_.

I see a great number of benefits to using a URI since this is what URI's are for, identification and only a very weak argument against.

@ianbjacobs said in #34:

> short strings for common methods, that are listed in the spec. The "formal" spec can list the "approved" ones and the Editor's draft can be updated easily for the "next ones to be approved but aren't yet."

I agree that historically systems that use URIs as identifiers have sometimes faced issues with developer adoption but I'd argue that this is more often the case when the system tries to accommodate lazy developers by making the whole system more complicated with aliases and short-names.

A developer that sees a payment request example as her first exposure to this standard and sees a list of payment methods that is a mix of short-strings and URIs would be understandably confused as to what the payment method identifier is. On the contrary, if these were all URIs and the majority resolved to a resource that helped them understand how to implement the payment method I'd consider that more developer friendly than short-string identifiers for a subset of the payment methods.

Further, a dependency on these URIs resolving to something useful and machine-readable is the shortcoming of many other systems that use URIs as identifiers yet in this system that would not be a hard requirement and the URi would still be useful long after it stopped resolving.

TL;DR: despite URIs being exactly the correct thing to use as identifiers on the Web (that is what they were designed for) there always seems to be resistance within the W3C to use them as such. In this instance the many reasons they have been problematic in the past are also not valid and so I would strongly encourage their use.

**Comment by dlongley** [comment]
I agree with @adrianhopebailie's reasoning. I think a registry is an unnecessary complexity, developers are more likely to do the right thing when there's one way to do it (and I believe they are much more likely to cut and paste than to type), and URIs are just the right tool here for a globally unique identifier.

**Comment by adrianba** [comment]
> This is starting to feel a lot like a duplicate of #34, should we close one of them?

They are two different issues:
* Issue #34 is about what form distributed extensibility should take. I'm hearing strong arguments for using a URL.
* Issue #35 (this issue) is about whether we allow well-known strings that aren't part of the distributed extensibility model to be recognized by the system.

The discussion of the two issues is muddled together but they are independent choices and closing one issue doesn't resolve the other so we should keep both.

**Comment by adrianhopebailie** [comment]
> Issue #34 is about what form distributed extensibility should take. I'm hearing strong arguments for using a URL.

:+1:  to URLs

> Issue #35 (this issue) is about whether we allow well-known strings that aren't part of the distributed extensibility model to be recognized by the system.

:-1: to the added complexity of mixing non-URL strings into the system when there doesn't appear to be any significant benefit beyond convenience and there are a host of potential problems this introduces

**Comment by justmoon** [comment]
I'm going to weigh in on the side of well-known strings. The idea is that you would only have a very limited number of them for the most widely recognized payment methods. I wouldn't underestimate the effect that elegance has on the adoption of a standard. It makes a difference when developers look at our code examples and go: "This looks really clean."

Also I dislike the idea of tying widely supported standards to domain names. Whoever owns the domain for a widely supported standardized payment method today, I don't know whether that'll still be the case ten years from now and what that URI will dereference to, so for payments methods that are actual published specs/recommendations from a major standards body, it seems safer to have a well-known string. For that (limited) case, the ability to dereference actually **introduces** issues of ensuring continuity in the ownership and operation of that domain etc. that can be avoided by using a neutral string.

Lastly if I think of other standards, this pattern seems to have worked well in the past. I'm happy that I can write `<link rel="stylesheet" .../>` instead of `<link rel="http://www.w3.org/2010/rels/stylesheet" .../>` or some such. Yes, I realize the latter may be the The Right Thing To Do, but for well-known values that are perhaps even part of the original standard it makes way more sense to use well-known strings.

Standards should be designed for humans (developers) not for machines. Humans will say "stylesheet" and (correctly and quite reasonably) expect you to know what they mean. That was the problem with the whole XML debacle in that it was very rigorous and well intentioned, but ultimately not designed for humans.

**Comment by ianbjacobs** [comment]
@justmoon,

Do you mean "well-known strings and not any other mechanism"? Or would you be ok with well-known strings plus URLs as the extension mechanism (until the next round of strings is blessed)?

Ian

**Comment by webpayments** [comment]
I would be fine with "well known strings" that are mapped (internally) to
URIs.
Citing your example above, rel="stylesheet" actually expands to rel="
http://www.w3.org/1999/xhtml/vocab#stylesheet" [1] [2].  No one knows that
nor cares, but certain processors do that expansion internally.  In RDFa
this is handled using something called an "initial context".  Basically it
is a set of rules that are well known and that implementations support.  It
allows the use of "well known strings" and also permits expansion.


[1] http://www.w3.org/1999/xhtml/vocab
[2] http://www.w3.org/TR/rdfa-core

On Wed, Jan 6, 2016 at 7:37 AM, ianbjacobs <notifications@github.com> wrote:

> @justmoon <https://github.com/justmoon>,
>
> Do you mean "well-known strings and not any other mechanism"? Or would you
> be ok with well-known strings plus URLs as the extension mechanism (until
> the next round of strings is blessed)?
>
> Ian
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/WICG/paymentrequest/issues/35#issuecomment-169324146>.
>



-- 
-Shane

**Comment by adrianhopebailie** [comment]
That's two pretty good arguments in favour of well-known strings as aliases for URIs.
The question then is how the registry of well-known strings should work?

**Comment by webpayments** [comment]
Referencing RDFa (again) as a potential model, it basically says that
mappings should be in at least one of N approved formats, where those
formats are well defined by RDFa. We could surely do the same. Have the
registry at a well known URI maintained by the W3C in one or more known,
stable formats (JSON, JSON-LD, HTML+RDFa for human readable).

The advantage of this approach is that it changes rarely.  The disadvantage
is that it requires some sort of consensus process to change the document.
In the case of RDFa and its "initial context" the W3C undertakes to update
it as new prefixes (the RDFa equivalent of a well known string) become
popular.


On Wed, Jan 6, 2016 at 7:58 AM, Adrian Hope-Bailie <notifications@github.com
> wrote:

> That's two pretty good arguments in favour of well-known strings as
> aliases for URIs.
> The question then is how the registry of well-known strings should work?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/WICG/paymentrequest/issues/35#issuecomment-169329438>.
>



-- 
-Shane

**Comment by msporny** [comment]
To build on what Shane is saying - this doesn't have to be an "either this or that" decision. We can easily have both (and this is what the Web Payments CG proposals have been suggesting for 4+ years now).

If you use a data format that supports "terms" that map to URLs (like JSON-LD), you can easily do both. With the example of JSON-LD in mind, do this in the JSON-LD Context:
```javascript
{ 
...
"paymentInstrumentType": {
  "@id": "https://w3c.org/ns/webpayments/v1#paymentInstrumentType",
  "@type": "@id"
}
"Visa": "https://visa.com/vocab#Visa"
...
}
```

and then your JSON data ends up looking like this:

```javascript
{
...
  "paymentInstrumentType": "Visa",
...
}
```

So, you get the three desired benefits: 
  * easily readable for developers
  * completely unambiguous for machines
  * decentralized innovation (Visa gets to own their terms, standards orgs gets to own their terms, each can innovate at their own pace)

Here's a link to a working example in the JSON-LD playgrond: http://json-ld.org/playground/#/gist/51e1b98df0906c2f793c

**Comment by msporny** [comment]
> The question then is how the registry of well-known strings should work?

The registry of well-known strings is just a JSON-LD context (if we decide to go that route, and this is really something we should decide early on because a large number of the problems we're discussing are solved if we go with many variations of that route).

I know a few folks are opposed to W3C managing a registry of some kind but if we want to do this sort of aliasing, the simple fact of the matter is that /some organization/ is going to have to manage the mapping. Even if we didn't want to map strings to URLs, some organization would still have to maintain the registry of "short strings that map to payment instruments".

Barring just purely using URLs (which I don't think anyone is arguing for), we're in the position of maintaining a registry of some kind. W3C and IETF has a long history of doing this, and there are examples of it working out well.

**Comment by ianbjacobs** [comment]
@msporny,

If we go with a registry of short strings, then I would make the specification the registry of approved strings. An editors draft could be the registry of "strings under consideration but not yet approved" and those could become "approved" in subsequent spec revisions.

I am suggesting we consider such an approach as a way of having an institutional commitment (via the WG) to maintaining the registry. While there are advantages to having a registry exist outside the spec (e.g., it could be updated beyond the life of a WG) at least for the time being I would propose we give the WG the responsibility of maintaining any registry.

Personally,  I think our approach should allow for both short strings and URIs, though I do not yet have a deep understanding of the cost of the added complexity.

Ian

**Comment by webpayments** [comment]
I am open to having the spec be the registry ONLY IF there is a way to
machine process it.  I should not have to visually inspect the document in
order to extract the collection of well known strings and their mappings.
Nor should I need to do this each time the spec is updated to ensure my app
/ wallet / whatever is up to date.

RDFa triples would be okay for this.  A JSON-LD companion document that is
"part" of the spec would also be okay for this.

On Wed, Jan 6, 2016 at 9:58 AM, ianbjacobs <notifications@github.com> wrote:

> @msporny <https://github.com/msporny>,
>
> If we go with a registry of short strings, then I would make the
> specification the registry of approved strings. An editors draft could be
> the registry of "strings under consideration but not yet approved" and
> those could become "approved" in subsequent spec revisions.
>
> I am suggesting we consider such an approach as a way of having an
> institutional commitment (via the WG) to maintaining the registry. While
> there are advantages to having a registry exist outside the spec (e.g., it
> could be updated beyond the life of a WG) at least for the time being I
> would propose we give the WG the responsibility of maintaining any registry.
>
> Personally, I think our approach should allow for both short strings and
> URIs, though I do not yet have a deep understanding of the cost of the
> added complexity.
>
> Ian
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/WICG/paymentrequest/issues/35#issuecomment-169367681>.
>



-- 
-Shane

**Comment by msporny** [comment]
Here's an example of how we could do the registry for all terms, short strings, etc. that would go in payment messages (human readable and machine-readable as RDFa) - the link is live, go to it:

https://w3id.org/payments

and the machine-readable version of the registry (as JSON-LD), again, live link:

https://w3id.org/payments/v1

The example registry above:
 * Uses the w3id.org redirect service as a best practice - the final redirect location for the registry may change from time to time, but the links above could be used for the next 50+ years due to the w3id.org site's continuity plan.
 * Resolves to the Editor's Draft for the human-readable spec and web-payments.org for the machine-readable one. We'd update these once we hit production with v1 of the vocabulary and context.

These rules could be used for updating the human-readable vocabulary:
 * Adding new terms can be done by consensus in the WPWG and then the WPIG (if the WPWG dissolves). Terms would start at 'unstable', moving to 'proposed', then 'stable', and if terms fall out of favor 'deprecated'. We would probably never remove terms.

The machine-readable context would be updated like so:
 * The context is versioned by major number (v1, v2, v3, etc.)
 * Adding terms requires consensus and a 'stable' marker in the human-readable vocabulary.
 * Terms are never changed or removed once they are added to a particular context version (doing so would break deployed systems). 
 * Terms may be removed as you move from v1 to v2, etc.
 * Nothing would prevent developers from using full URLs where terms, like "Visa" are used.

None of what's said above wrt. the management of these registries should be controversial. They reflect best practices that have been refined over the past 10+ years in the Linked Data realm at W3C. As Shane said, we used many of these rules to manage the RDFa initial context and that's worked out pretty well so far (as far as managing the registry is concerned).

**Comment by ianbjacobs** [comment]
@msporny,

I recognize the value of grounding the meaning of the terms. However, the argument has
been made in other contexts that prefixing mechanisms that require developers to put context explicitly in code as a separate URI can make the code more brittle (e.g., in the face of cut and paste). 

Can we make the grounding implicit? The API overall would be grounded (either to a specific version of the API if that's what's needed, or a latest version potentially if the application prefers that approach) and that's where one would find the definitions of the short strings. We would avoid changing the meaning of a short string over time (at least in ways that would lead to incompatibilities in deployed software). Thus, we could use 'visa2' to signal a significant change from the meaning of 'visa'. This would simplify life for developers and avoid the confusion and cost of relying on dereferencing a URI to establish that identical strings in fact mean different things.

In exchange for this reduction in flexibility, I still favor the ability to use either a short string or a URL to identify a payment app. I recognize that this option increases the processing cost for implementers.

Ian

**Comment by msporny** [comment]
> Can we make the grounding implicit?

I think doing that would be a bad idea as you end up, in time, w/ certain organizations mis-implementing the basic context and inevitably this will lead to an interoperability problem when one large multinational starts using "currency" to mean something slightly different from another large multinational. We should make the processing rules explicit, and the best way to do that would be to say that developers SHOULD provide a context in all of their messages. If we do that, there is no chance that messages could be misinterpreted.

We could also say in the spec that in the absence of a context, all Web Payments messages MUST assume a context of "https://w3id.org/payments/v1". This is, of course, a compromise and I hope the group wouldn't make this the recommendation for developers.

Note that including the context should have no effect on the browser API implementations since they are to pass through the data untouched.

**Comment by msporny** [comment]
> The API overall would be grounded ... to a latest version potentially if the application prefers that approach

This would be a disaster if a term ever changed it's range (allowable values) or semantic meaning (and we have examples of where that has happened in the past).

You can't have a message that is sent from a system and be consumed by two applications w/o having ONE clear mechanism for interpreting that message. If a v1 system sends the same message to a v2 system and a v3 system, the message could be interpreted by the v2 and v3 system differently (to catastrophic effect, since these are payment systems). Similarly, a v2 and v3 response without a known context being processed by a v1 system could result in something similarly terrible happening.

If you're not going to ground the message to a context, you have to ground the message to a version. 

A context grounds a message to a version explicitly and formally.

A version number grounds a message to a version explicitly, but informally (the message could still be interpreted by two systems differently, especially when the payload of the message can be extended by the sender).

**Comment by msporny** [comment]
> avoid the confusion and cost of relying on dereferencing a URI to establish that identical strings in fact mean different things.

You don't have to dereference the URI. Systems can and do hard-code interpretation based on well-known context URLs and clear update methodologies on those contexts.

If the URL is well known and the data at that URL is only ever added to, hard coding or long-term caching of the context in your application is something that we'd expect people to do.

The key here is that we have some formal definition with provable correctness vs. "just a list of words that a bunch of yahoos in a WG have slapped together and put in a human readable document online". :)

**Comment by ianbjacobs** [comment]
@msporny 

> I think doing that would be a bad idea as you end up, in time, w/ certain organizations is-implementing the basic context 

I'm not sure why that follows. 

> We should make the processing rules explicit,

We would, via the specification of the API.

> We could also say in the spec that in the absence of a context, all Web Payments messages MUST 
> assume a context of "https://w3id.org/payments/v1". This is, of course, a compromise and I hope the 
> group wouldn't make this the recommendation for developers.

I understand that approach, but in the end, it amounts to providing two pieces of information (a URL and a short string that's separate from it). It seems simpler that when you want to extend beyond the specificaiton, you provide just a single URL. It is possible that you could get back some data by deferencing that URL, but that would be behavior specific to that extension.

> This would be a disaster if a term ever changed it's range (allowable values) or semantic meaning 
> (and we have examples of where that has happened in the past).

I agree we should work hard not to create compatibility for a published short string. My suggestion is that when we need to break compatibility we mint a new short label.

> You can't have a message that is sent from a system and be consumed by two 
> applications w/o having ONE clear mechanism for interpreting that message.

Where versioning is needed, it should be done at the level of the whole API. I don't think the messages themselves have meaning outside of SOME API (and we are likely to define more than one of them).

> The key here is that we have some formal definition with provable correctness vs.
> "just a list of words that a bunch of yahoos in a WG have slapped together and put in a human 
> readable document online". :)

I agree we both want that. I am proposing that the WG publish that list via the specification, and it is via the API that one identifies the specification. (I am assuming the API identifies its origin specification.)

Ian

**Comment by msporny** [comment]
>> I think doing that would be a bad idea as you end up, in time, w/ certain organizations is-implementing the basic context

> I'm not sure why that follows.

You said make the grounding implicit. When you don't make a grounding explicit, you leave a developer with room to add extra bits in that could be problematic as the system evolves.

That doesn't always follow, but it sometimes does and those "sometimes" cases can cause fairly significant interoperability problems.

For example - do we allow or prevent organizations from putting a (today) non-standard "referenceId" into a payment message? If we don't explicitly ground the context, we definitely allow for that. What happens when a developer adds in something that we don't account for in our "implicit" grounding?

I think many of us who desire the basic message format to be extensible, would want to allow a company to put in a referenceId if it solved a business problem for them (even though the specification does not say that 'referenceId' is a known term). We should, however, make the grounding of that term explicit - how does the ecosystem process 'referenceId'? We should have an answer for that question, and my concern is that an implicit grounding doesn't answer that question.

Now, consider in v2 that we think 'referenceId' is a good idea (without knowing that other applications have started using it). If we go with an implicit grounding, and we launch v2, then all of a sudden we have a problem. Old systems w/ old messages w/ referenceId will now conflict w/ new systems w/ old messages (again, because the grounding is implicit, there is no way to tell an old message from a new message).

**Comment by msporny** [comment]
>> We should make the processing rules explicit,

> We would, via the specification of the API.

How would we do that (specifically)?

> Where versioning is needed, it should be done at the level of the whole API.

How is versioning at the API level done?

>  I don't think the messages themselves have meaning outside of SOME API (and we are likely to define more than one of them).

That we're likely to define more than one API is exactly why the messages have meaning outside of an API. Data has meaning outside of an API.

**Comment by ianbjacobs** [comment]
> For example - do we allow or prevent organizations from putting a (today) non-standard "referenceId" > into a payment message? If we don't explicitly ground the context, we definitely allow for that. What 
> happens when a developer adds in something that we don't account for in our "implicit" grounding?

Behavior is undefined. It's the same with explicit grounding and unknown short labels.

> I think many of us who desire the basic message format to be extensible, would want to allow a 
> company to put in a referenceId if it solved a business problem for them (even though the specification > does not say that 'referenceId' is a known term). We should, however, make the grounding of that term > explicit - how does the ecosystem process 'referenceId'? We should have an answer for that question, > and my concern is that an implicit grounding doesn't answer that question.

My suggested extensibility mechanism is a single URI. I don't think we should optimize for identifier length/simplicity in the extension case. So people can extend as needed with a full URI and provide whatever applications need to know at that URI. 

So the implicit grounding is only for WG-approved short strings, and those are defined in the specification referenced by the API.

> How is versioning at the API level done?

I don't have any suggestions here (because I am not an API designer). I would look to see how other APIs manage versioning as I imagine that's not a new problem.

> That we're likely to define more than one API is exactly why 
> the messages have meaning outside of an API. Data has meaning outside of an API.

I chose my phrase poorly and appreciate your reply. So I'll rephrase: 

* I personally don't think it is high-priority design goal that the data be reusable outside of some API that we are designing (whether JS or HTTP). While in the abstract I see the value (1) I don't think we've discussed a concrete use case that is driving this and (2) there is a cost in terms of verbosity and brittleness.
* If we can factor out the message data to make it easier to design our APIs, that seems useful (but not mandatory).

This sounds like a question to put to the group: what is the priority that the message data be designed to be reusable (meaningful, grounded, etc.) outside of the context of some API defined by the group?

Ian

**Comment by msporny** [comment]
> This sounds like a question to put to the group: what is the priority that the message data be designed to be reusable (meaningful, grounded, etc.) outside of the context of some API defined by the group?

Looks like this issue is getting broader than the topic that was raised (I'm partly to blame for that). So, trying to re-focus on the original topic:

> Should well-known identifiers be used for ubiquitous payment methods

I think we have consensus on:

PROPOSAL: Well-known identifiers like "Visa" and "MasterCard" should be used for ubiquitous payment methods.
PROPOSAL: The extensibility mechanism for payment method identifiers is the URL.

We don't yet know if:

1. The well-known identifiers registry lives in the Web Payments specification.
2. The well-known identifiers registry should be machine readable.
3. There is a formal mechanism for expressing the payment method registry (like a JSON-LD context).
4. The APIs are versioned (somehow).
5. The behavior of unknown payment message data is undefined.
6. Payment message data should be designed to have meaning outside of the context of an API call defined by the group.

... and I'm sure there are other unknowns that this thread has identified.

**Comment by dlongley** [comment]
I realize "Visa" and "MasterCard" are just examples, but this is just a reminder that actual Payment Method identifiers are likely to be more specific than that, for example, they may be more like "Legacy Visa" or "Some Special MasterCard Scheme", see: https://github.com/w3c/webpayments/issues/15#issuecomment-162527538

**Comment by ianbjacobs** [comment]
@msporny,

Regarding the two proposals, yes, that would be my preference but I believe the group likely needs more information on the cost of supporting both short strings and URLs.

Regarding your list of six points:

* Versioning. This sounds useful, but I'd like to hear from people with more API design experience (and on how it is done).
* Machine readability. I can see a case for having a machine-readable expression of short strings (e.g., validity checking of some sort). We could discuss various ways to publish the data (in the spec, with the spec).
* Nature of registry. As I wrote, I have a mild preference for having the spec (including any machine readable info) be the registry while we have an institutional commitment through the WG. But I understand the case for making it easier to update the registry through some process.
* Regarding "The behavior of unknown payment message data is undefined." I would phrase it this way:
If we support short strings, then if implementations of the API encounter undefined short names, we should be sure to address error handling in the API specs.
* Regarding "There is a formal mechanism for expressing the payment method registry (like a JSON-LD context).": the specification would be the formal mechanism. So I think this bullet goes away in favor of the previous one on machine-readability.
* Regarding "Payment message data should be designed to have meaning outside of the context of an API call defined by the group.". I think that's close to the question, but perhaps the key point has to do with the data being self-defining (is that the phrase?) via URLs that enable discovery of the context where the data are defined?

**Comment by adrianba** [comment]
Now tracked by w3c/browser-payment-api#10.
